File: app/src/main/java/com/myflightbook/android/DlgDatePicker.java
/*
	MyFlightbook for Android - provides native access to MyFlightbook
	pilot's logbook
    Copyright (C) 2017-2020 MyFlightbook, LLC

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.myflightbook.android;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.widget.DatePicker;
import android.widget.DatePicker.OnDateChangedListener;
import android.widget.TimePicker;
import android.widget.TimePicker.OnTimeChangedListener;

import com.myflightbook.android.WebServices.UTCDate;

import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

class DlgDatePicker extends Dialog implements android.view.View.OnClickListener, OnDateChangedListener, OnTimeChangedListener {

    interface DateTimeUpdate {
        void updateDate(int id, Date dt);
    }

    enum datePickMode {LOCALDATEONLY, LOCALDATENULLABLE, UTCDATETIME}

    private Date m_Date = new Date();
    private datePickMode m_Mode;
    int m_id;    // the id of the date/time to update

    static boolean fUseLocalTime = false;

    DateTimeUpdate m_delegate = null;

    private DatePicker dpDate;
    private TimePicker tpTime;

    private void Init(datePickMode dpm, Date dt) {
        setContentView(R.layout.datepicker);
        setMode(dpm);
        setDate(dt);
    }

    DlgDatePicker(Context c, datePickMode dpm, Date dt) {
        super(c, R.style.MFBDialog);
        Init(dpm, dt);
    }

    public void onCreate(Bundle savedInstanceState) {
        findViewById(R.id.btnOK).setOnClickListener(this);
        findViewById(R.id.btnDateNow).setOnClickListener(this);
        findViewById(R.id.btnDateNone).setOnClickListener(this);

        GregorianCalendar c = getCalendar();
        c.setTime(m_Date);

        dpDate = findViewById(R.id.datePicker);
        tpTime = findViewById(R.id.timePicker);

        int y = c.get(Calendar.YEAR);
        int m = c.get(Calendar.MONTH);
        int d = c.get(Calendar.DAY_OF_MONTH);
        int h = c.get(Calendar.HOUR_OF_DAY);
        int min = c.get(Calendar.MINUTE);

        dpDate.init(y, m, d, this);
        tpTime.setOnTimeChangedListener(this);
        tpTime.setIs24HourView(true);
        tpTime.setHour(h);
        tpTime.setMinute(min);
    }

    private void updatePickers(Date dt) {
        GregorianCalendar c = getCalendar();
        c.setTime(dt);

        int y = c.get(Calendar.YEAR);
        int m = c.get(Calendar.MONTH);
        int d = c.get(Calendar.DAY_OF_MONTH);
        int h = c.get(Calendar.HOUR_OF_DAY);
        int min = c.get(Calendar.MINUTE);

        if (dpDate != null)
            dpDate.updateDate(y, m, d);
        if (tpTime != null) {
            tpTime.setHour(h);
            tpTime.setMinute(min);
        }
    }

    private void setDate(Date dt) {
        updatePickers(m_Date = dt);
    }

    public Date getDate() {
        return m_Date;
    }

    private void setMode(datePickMode dp) {
        m_Mode = dp;
        findViewById(R.id.timePicker).setVisibility((dp == datePickMode.UTCDATETIME) ? View.VISIBLE : View.GONE);
        findViewById(R.id.btnDateNone).setVisibility((dp == datePickMode.LOCALDATEONLY) ? View.GONE : View.VISIBLE);
    }

    private void NotifyDelegate() {
        if (this.m_delegate != null)
            m_delegate.updateDate(m_id, m_Date);
    }

    public void onClick(View v) {
        int id = v.getId();
        if (id == R.id.btnDateNone) {
            m_Date = UTCDate.NullDate();
            NotifyDelegate();
            dismiss();
        } else if (id == R.id.btnDateNow) {
            setDate(new Date());
            NotifyDelegate();
        } else if (id == R.id.btnOK)
            dismiss();
    }

    private GregorianCalendar getCalendar() {
        GregorianCalendar c;
        if (fUseLocalTime || m_Mode == datePickMode.LOCALDATEONLY || m_Mode == datePickMode.LOCALDATENULLABLE)
            c = new GregorianCalendar();
        else
            c = new GregorianCalendar(UTCDate.getUTCTimeZone());

        return c;
    }

    public void onDateChanged(DatePicker vw, int year, int monthOfYear, int dayOfMonth) {
        GregorianCalendar c = getCalendar();
        c.set(year, monthOfYear, dayOfMonth, tpTime.getHour(), tpTime.getMinute());

        this.m_Date = c.getTime();

        NotifyDelegate();
    }

    public void onTimeChanged(TimePicker view, int hourOfDay, int minute) {
        GregorianCalendar c = getCalendar();

        c.set(dpDate.getYear(), dpDate.getMonth(), dpDate.getDayOfMonth(), hourOfDay, minute);
        this.m_Date = c.getTime();

        if (m_Mode == datePickMode.UTCDATETIME)
            NotifyDelegate();
    }
}

File: app/src/main/java/com/myflightbook/android/DlgDatePicker.kt
/*
	MyFlightbook for Android - provides native access to MyFlightbook
	pilot's logbook
    Copyright (C) 2017-2022 MyFlightbook, LLC

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.myflightbook.android

import android.app.Dialog
import android.content.Context
import com.myflightbook.android.webservices.UTCDate.getNullDate
import com.myflightbook.android.webservices.UTCDate.getUTCTimeZone
import android.widget.DatePicker.OnDateChangedListener
import android.widget.TimePicker.OnTimeChangedListener
import android.widget.DatePicker
import android.widget.TimePicker
import android.os.Bundle
import android.view.View
import java.util.*

class DlgDatePicker(c: Context?, dpm: DatePickMode, dt: Date) : Dialog(
    c!!, R.style.MFBDialog
), View.OnClickListener, OnDateChangedListener, OnTimeChangedListener {
    interface DateTimeUpdate {
        fun updateDate(id: Int, dt: Date?)
    }

    enum class DatePickMode {
        LOCALDATEONLY, LOCALDATENULLABLE, UTCDATETIME
    }

    private var mDate = Date()
    private var mMode: DatePickMode? = null
    @JvmField
    var mId // the id of the date/time to update
            = 0
    @JvmField
    var mDelegate: DateTimeUpdate? = null
    private var dpDate: DatePicker? = null
    private var tpTime: TimePicker? = null
    private fun init(dpm: DatePickMode, dt: Date) {
        setContentView(R.layout.datepicker)
        setMode(dpm)
        date = dt
        setCanceledOnTouchOutside(false)
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        findViewById<View>(R.id.btnOK).setOnClickListener(this)
        findViewById<View>(R.id.btnDateNow).setOnClickListener(this)
        findViewById<View>(R.id.btnDateNone).setOnClickListener(this)
        val c = calendar
        c.time = mDate
        dpDate = findViewById(R.id.datePicker)
        tpTime = findViewById(R.id.timePicker)
        val y = c[Calendar.YEAR]
        val m = c[Calendar.MONTH]
        val d = c[Calendar.DAY_OF_MONTH]
        val h = c[Calendar.HOUR_OF_DAY]
        val min = c[Calendar.MINUTE]
        val vwDate = dpDate
        val vwTime = tpTime
        vwDate!!.init(y, m, d, this)
        vwTime!!.setOnTimeChangedListener(this)
        vwTime.setIs24HourView(true)
        vwTime.hour = h
        vwTime.minute = min
    }

    private fun updatePickers(dt: Date) {
        val c = calendar
        c.time = dt
        val y = c[Calendar.YEAR]
        val m = c[Calendar.MONTH]
        val d = c[Calendar.DAY_OF_MONTH]
        val h = c[Calendar.HOUR_OF_DAY]
        val min = c[Calendar.MINUTE]
        if (dpDate != null) dpDate!!.updateDate(y, m, d)
        if (tpTime != null) {
            tpTime!!.hour = h
            tpTime!!.minute = min
        }
    }

    var date: Date
        get() = mDate
        private set(dt) {
            updatePickers(dt.also { mDate = it })
        }

    private fun setMode(dp: DatePickMode) {
        mMode = dp
        findViewById<View>(R.id.timePicker).visibility =
            if (dp == DatePickMode.UTCDATETIME) View.VISIBLE else View.GONE
        findViewById<View>(R.id.btnDateNone).visibility =
            if (dp == DatePickMode.LOCALDATEONLY) View.GONE else View.VISIBLE
    }

    private fun notifyDelegate() {
        if (mDelegate != null) mDelegate!!.updateDate(mId, mDate)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btnDateNone -> {
                mDate = getNullDate()
                notifyDelegate()
                dismiss()
            }
            R.id.btnDateNow -> {
                date = Date()
                notifyDelegate()
            }
            R.id.btnOK -> {
                // Issue #233 - typing and then hitting OK without committing what you typed loses typed changes.
                val tp = findViewById<TimePicker>(R.id.timePicker)
                val dp = findViewById<DatePicker>(R.id.datePicker)
                tp.clearFocus()
                dp.clearFocus()
                notifyDelegate()
                dismiss()
            }
        }
    }

    private val calendar: GregorianCalendar
        get() {
            return if (fUseLocalTime || mMode == DatePickMode.LOCALDATEONLY || mMode == DatePickMode.LOCALDATENULLABLE) GregorianCalendar() else GregorianCalendar(
                getUTCTimeZone
            )
        }

    override fun onDateChanged(vw: DatePicker, year: Int, monthOfYear: Int, dayOfMonth: Int) {
        val c = calendar
        c[year, monthOfYear, dayOfMonth, tpTime!!.hour, tpTime!!.minute] = 0
        mDate = c.time
    }

    override fun onTimeChanged(view: TimePicker, hourOfDay: Int, minute: Int) {
        val c = calendar
        c[dpDate!!.year, dpDate!!.month, dpDate!!.dayOfMonth, hourOfDay, minute] = 0
        mDate = c.time
    }

    companion object {
        @JvmField
        var fUseLocalTime = false
    }

    init {
        init(dpm, dt)
    }
}